src/main/java/org/betonquest/betonquest/quest/event/stage/StageEventFactory.java

```diff
@@ -0,0 +1,85 @@
+package org.betonquest.betonquest.quest.event.stage;
+
+import org.betonquest.betonquest.BetonQuest;
+import org.betonquest.betonquest.Instruction;
+import org.betonquest.betonquest.VariableNumber;
+import org.betonquest.betonquest.VariableString;
+import org.betonquest.betonquest.api.profiles.Profile;
+import org.betonquest.betonquest.api.quest.event.Event;
+import org.betonquest.betonquest.api.quest.event.EventFactory;
+import org.betonquest.betonquest.exceptions.InstructionParseException;
+import org.betonquest.betonquest.exceptions.QuestRuntimeException;
+import org.betonquest.betonquest.id.ObjectiveID;
+import org.betonquest.betonquest.objectives.StageObjective;
+
+import java.util.Locale;
+
+/**
+ * Factory to create stage events to modify a StageObjective.
+ */
+public class StageEventFactory implements EventFactory {
+    /**
+     * BetonQuest instance.
+     */
+    private final BetonQuest betonQuest;
+
+    /**
+     * Create the stage event factory.
+     *
+     * @param betonQuest BetonQuest instance
+     */
+    public StageEventFactory(final BetonQuest betonQuest) {
+        this.betonQuest = betonQuest;
+    }
+
+    @Override
+    public Event parseEvent(final Instruction instruction) throws InstructionParseException {
+        final ObjectiveID objectiveID;
+        try {
+            objectiveID = instruction.getObjective();
+        } catch (final ClassCastException e) {
+            throw new InstructionParseException("Objective '" + instruction.getObjective() + "' is not a stage objective", e);
```

Comment:
This does not throw a `ClassCastException`. It throws an `InstructionParseException`.
Also, you cannot call `instruction.getObjective()` again - this will call `instruction.next()` internally and will cause another exception. Call `instruction.next()` yourself, save it to a variable and then try to obtain the objective. If this causes an exception use the saved result to log the error.